### PR TITLE
libavfilter: add cuda signature filter

### PR DESCRIPTION
**Pull request by oscar-davids**
CUDA implementation of MPEG-7 video signature filter #11
 

- [X] implementation of signature for GPU
- [X] update document
- [X] test and benchmarking on single lane PCI GPU.

### CODE REVIEW

**Review by jailuthra — COMMENTED** [review]
Nice, testing this by sending individual segments and comparing their signatures using the LPMS API works fine. But I have a few questions and suggestions -

1. What are the performance and accuracy metrics of `signature_cuda` filter compared to `signature` filter? I wasn't able to test it via the [script](https://gist.github.com/yondonfu/084eb52062cc12a54f5d737174d1d389) as the new filter doesn't support multiple inputs, or print out matched frames.

2. A lot of code that is copied over from `signature` filter into the `signature_cuda` filter, making it hard to review what exactly changed between the copying and why. While this does work end-to-end, it also makes maintaining both filters tricky, as any changes we (or upstream) makes to the `signature` filter need to be copied over to `signature_cuda` filter.

Would it be possible to move the CUDA scaling kernel as a separate module (similar to how `signature_lookup.c` is a module but not a filter). Both CPU and GPU scalers can then be called via a single `signature` filter depending on a `scaling_backend=cuda` AVOption? This would address both (1) and (2). WDYT?

**Comment by oscar-davids** [comment]
I think the mixing of CUDA and CPU versions is not good because it would be failing in the compiling if there's no GPU. 
As we know, the "signature_cuda" filter or module has dependent on the Cuda library.
And I moved all duplicated functions to the signature_lookup.c file. In the future, I am going to consider comparing multi-files using the GPU decoder.  [Here ](https://github.com/livepeer/FFmpeg/issues/11#issuecomment-922343961)are the comparative result(signature_cuda vs signature) and script.

File: configure [hunk 1]
```diff
@@ -4352,10 +4354,10 @@ fi
 
 if enabled cuda_nvcc; then
     nvcc_default="nvcc"
-    nvccflags_default="-gencode arch=compute_52,code=sm_52 -O2"
+    nvccflags_default="-gencode arch=compute_35,code=sm_35 -O2"
```
Comment (jailuthra) [inline configure]: I believe we switched from `sm_30` -> `sm_52` to support cuda 11.4. From what I understand `sm_35` is also for Kepler generation of GPUs and is deprecated from cuda 11 but not completely removed. Any particular reason we're switching to it?

**Review by jailuthra — APPROVED** [review]
LGTM!

For the short-term the solution of moving the common parts in `signature_loookup.c` sounds good to me. Might be problematic in future if there are changes upstream around the shared code, and we want to rebase.

File: configure (hunk 1 above)
Comment (oscar-davids) [reply configure]: There is no special reason and I just modified it for compiling in my environment(cuda sdk 11.2, K80), I think it is related to GPU compute capability. I will restore it.

**Pull request merged**